File: src/duJavaAPI/Navigator.java
/*    
Navigator.java 
Copyright (C) 2020 Stephane Boivin (Devgeek studio enr.)

This file is part of "DU offline sandbox API".

"DU offline sandbox API" is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

"DU offline sandbox API" is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with "DU offline sandbox API".  If not, see <https://www.gnu.org/licenses/>.
*/
package duJavaAPI;

import java.awt.Color;

import javax.swing.ImageIcon;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.border.LineBorder;

import org.luaj.vm2.LuaTable;
import org.luaj.vm2.lib.jse.CoerceJavaToLua;

import chrriis.dj.nativeswing.swtimpl.components.JWebBrowser;
import offlineEditor.DUElement;
import offlineEditor.execTimer;

public class Navigator extends BaseElement {
public JWebBrowser web = null;
public String html = "";
int sizeX = 106;
int sizeY = 134;   
int timerId = 0;
int isRemoteControlled = 0;

public Navigator(int pid, int px, int py, boolean pverboseJava) {
	name = "Navigator";
	x = px;
	y = py;		
	
	verboseJava = pverboseJava;

	panel = new JPanel();		
	panel.setLayout(null);
	// panel.setVisible(false);
	panel.setBounds(px, py, sizeX, sizeY+16);
	panel.setBorder(LineBorder.createBlackLineBorder());
	panel.setBackground(Color.black);

	// label
	JLabel  lblname = new JLabel(name);
	lblname.setForeground(Color.white);
	lblname.setBounds(8, 2, sizeX, 16);	
	panel.add(lblname, 1, 0);		

	// white frame
	JPanel lblPicWhite = new JPanel();		
 	lblPicWhite.setBounds(1,22,104,127);
 	lblPicWhite.setBackground(Color.white);
	panel.add(lblPicWhite, 1, 0);

 	// picture
 	JLabel lblPic = new JLabel();		
	lblPic.setBounds(2,22,100,130);
	lblPic.setIcon(new ImageIcon("src/pictures/elements/Navigator.png")); 
	panel.add(lblPic, 1, 0);

	// lua require and interface 
	urlAPI = "src/duElementAPI/navigator.lua";
    luaCall = name+" = createInterfaceNavigator(\""+pid+"\", \""+name+"\")";		
    if(verboseJava) System.out.println("[JAVA] Navigator created");

}

@Override	
public void update(String[] param) {
	String pcommand = param[0];
    String[] cmd1 = {"getType"};
    String[] cmd2 = {"getState"};    
    String[] cmd3 = {"0", " "};
//    System.out.println("[JAVA] Navigator update (" + param[0] + ", " + param[1] + ", " + param[2] + ")"); 
    switch (pcommand) {
		case "setEngineCommand": // to do
			// param[1] taglist 
			// param[2] acceleration (vec3)
			// param[3] angularAcceleration (vec3) 
			// param[4] keepForceCollinearity
			// param[5] keepTorqueCollinearity, 
			// param[6] priority1SubTags
			// param[7] priority2SubTags
			// param[8] priority3SubTags
			// param[9] toleranceRatioToStopCommand
			break;
		case "setEngineThrust": // to do
			// param[1] taglist 
			// param[2] thrust (newton)
		case "setAxisCommandValue": // to do
			// param[1] axis 
			// param[2] commandValue
			break;
		case "setupAxisCommandProperties": // to do
			// param[1] axis 
			// param[2] commandType
			break;
		case "cancelCurrentControlMasterMode": // to do
			break;
		case "setupControlMasterModeProperties": // to do
			// param[1] controlMasterModeId
			// param[2] displayName
			break;
		case "extendLandingGears":
			for (DUElement elem : sandbox.elements) {
			   if(elem == null) continue;
			   if(elem.GetType().equals("duJavaAPI.ToggleUnit")) {
				   if(elem.element.get(cmd1).equals("LandingGear") ) {
					   if((int)elem.element.get(cmd2) == 0) {
						   cmd3[0] = "1";
						   cmd3[1] = elem.element.name;
						   elem.element.update(cmd3);
					   }
				   }
			   }
			}
			break;
		case "retractLandingGears": 
			for (DUElement elem : sandbox.elements) {
			   if(elem == null) continue;
			   if(elem.GetType().equals("duJavaAPI.ToggleUnit")) {
				   if(elem.element.get(cmd1).equals("LandingGear") ) {
					   if((int)elem.element.get(cmd2) == 1) {
						   cmd3[0] = "0";
						   cmd3[1] = elem.element.name;
						   elem.element.update(cmd3);
					   }
				   }
			   }			   
			}
		    break;
		case "switchOnHeadlights": 
			for (DUElement elem : sandbox.elements) {
			   if(elem == null) continue;
			   if(elem.GetType().equals("duJavaAPI.ToggleUnit")) {
				   if(elem.element.get(cmd1).equals("Light") ) {
					   if((int)elem.element.get(cmd2) == 0) {
						   cmd3[0] = "1";
						   cmd3[1] = elem.element.name;
						   elem.element.update(cmd3);
					   }
				   }
			   }			   
			}
		    break;
		case "switchOffHeadlights": 
			for (DUElement elem : sandbox.elements) {
			   if(elem == null) continue;
			   if(elem.GetType().equals("duJavaAPI.ToggleUnit")) {
				   if(elem.element.get(cmd1).equals("Light") ) {
					   if((int)elem.element.get(cmd2) == 1) {
						   cmd3[0] = "0";
						   cmd3[1] = elem.element.name;
						   elem.element.update(cmd3);
					   }
				   }
			   }			   
			}
		    break;
		case "_setRemoteControlled": 
			isRemoteControlled = Integer.valueOf(param[1]);
			break;
    }
}

@Override
public Object get(String[] param) {    	
	String pcommand = param[0];
    switch (pcommand) {
		case "isRemoteControlled": 
			return isRemoteControlled;
		case "isMouseControlActivated": 
			return 1;
		case "getAtmosphereDensity": // to do 
	        return .9;
		case "getClosestPlanetInfluence": // to do
	        return 1;
		case "getAxisCommandValue": // to do
			// param[1] axis 
	        return 1;
		case "getControlMasterModeId": // to do
			return 1;
		case "isAnyLandingGearExtended": 
	        String[] cmd1 = {"getType"};
	        String[] cmd2 = {"getState"};
			for (DUElement elem : sandbox.elements) {
			   if(elem == null) continue;
			   if(elem.GetType().equals("duJavaAPI.ToggleUnit")) {
				   if(elem.element.get(cmd1).equals("LandingGear") ) {
					   if((int)elem.element.get(cmd2) == 1) return 1; 
				   }
			   }			   
			} 
			return 0;
    }    
    return -1;
 }	
}

File: src/duElementAPI/navigator.lua
--[[    
    navigator.lua 
    Copyright (C) 2020 Stephane Boivin (Discord: Nmare418#6397)
    
    This file is part of "DU offline sandbox API".

    "DU offline sandbox API" is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    "DU offline sandbox API" is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with "DU offline sandbox API".  If not, see <https://www.gnu.org/licenses/>.
]]

Inavigator = {id, name}

function Inavigator:new(o)
  o = o or {}
  setmetatable(o, self)
  self.__index = self

  o.maxHitPoints = 2000
  o.hitPoints = 2000
  o.mass = 34.06475
  o.class = 'CockpitUnit'
  o.integrity = 100
  
  o.navigator = Navigator:new{id = o.id, name = 'Navigator'}
  o.generic = genericElement:new{id = o.id, maxHitPoints = o.maxHitPoints, hitPoints = o.hitPoints, mass = o.mass, class = o.class, integrity = o.integrity }

  return o
end

function createInterfaceNavigator(id)
  local obj = Inavigator:new{id=id}

  obj.new = function(system, core, unit) return obj end -- notice: system, core and unit are already connected on the java engine side
  obj.getAtmosphereDensity = function() return obj.navigator:getAtmosphereDensity() end
  obj.getClosestPlanetInfluence = function() return obj.navigator:getClosestPlanetInfluence() end
  
  obj.setEngineCommand = function(taglist, acceleration, angularAcceleration, keepForceCollinearity, keepTorqueCollinearity, 
                                  priority1SubTags, priority2SubTags, priority3SubTags, toleranceRatioToStopCommand)
  obj.navigator:setEngineCommand(taglist, acceleration, angularAcceleration, keepForceCollinearity, keepTorqueCollinearity, 
                                   priority1SubTags, priority2SubTags, priority3SubTags, toleranceRatioToStopCommand)
  end         
  obj.setEngineThrust = function(taglist, thrust) obj.navigator:setEngineThrust(taglist, thrust) end
  obj.setAxisCommandValue = function(axis, commandValue) obj.navigator:setAxisCommandValue(axis, commandValue) end
  obj.getAxisCommandValue = function(axis) return obj.navigator:getAxisCommandValue(axis) end
  obj.setupAxisCommandProperties = function(axis, commandType) obj.navigator:setupAxisCommandProperties(axis, commandType) end
  obj.getControlMasterModeId = function() return obj.nagivator:getControlMasterModeId() end
  obj.cancelCurrentControlMasterMode = function() obj.navigator:cancelCurrentControlMasterMode() end
  obj.setupControlMasterModeProperties = function(controlMasterModeId, displayName) obj.navigator:setupControlMasterModeProperties(controlMasterModeId, displayName) end 
  obj.isAnyLandingGearExtended = function() return obj.navigator:isAnyLandingGearExtended() end
  obj.extendLandingGears = function() obj.navigator:extendLandingGears() end
  obj.isRemoteControlled = function() return obj.navigator:isRemoteControlled() end
  obj.isMouseControlActivated = function() return obj.navigator:isMouseControlActivated() end
  obj.switchOnHeadlights = function() obj.navigator:switchOnHeadlights() end
  obj.switchOffHeadlights = function() obj.navigator:switchOffHeadlights() end
  obj._setRemoteControlled = function(remoteControlled) obj.navigator:_setRemoteControlled(remoteControlled) end -- 0 false/1 true
  -- base elements
  obj.hide = function() obj.generic:hide() end
  obj.show = function() obj.generic:show() end
  obj.getData = function() obj.generic:getData() end
  obj.getId = function() return obj.generic:getId() end
  obj.getMass = function() return obj.generic:getMass() end
  obj.getElementClass = function() return obj.generic:getElementClass() end
  obj.getIntegrity = function() return obj.generic:getIntegrity() end
  obj.getHitPoints = function() return obj.generic:getHitPoints() end
  obj.getMaxHitPoints = function() return obj.generic:getMaxHitPoints() end
  obj._setMass = function(m) obj.generic:setMass(m) end
  obj._setMaxHitPoints = function(mhp) obj.generic:setMaxHitPoints(mhp) end
  obj._setHitPoints = function(hp) obj.generic:setHitPoints(hp) end
  obj._setIntegrity = function(int) obj.generic:setIntegrity(int) end

  return obj
end

Navigator = {id, name = "Navigator"}

function Navigator:new(o)
  o = o or {}  
  setmetatable(o, self)
  self.__index = self

  self.name = "Navigator"

  if _verboseLUA == 1 then print("[LUA] Navigator ["..o.name.."] created with the id ["..o.id.."]") end
    
  return o
end

function Navigator:_setRemoteControlled() 
  JavaWindow:setState(self.id, {"_setRemoteControlled"})   
end

function Navigator:isRemoteControlled() 
  return JavaWindow:get(self.id, {"isRemoteControlled"})   
end

function Navigator:isMouseControlActivated() 
  return JavaWindow:get(self.id, {"isMouseControlActivated"})   
end

function Navigator:switchOnHeadlights() 
  JavaWindow:setState(self.id, {"switchOnHeadlights"})   
end
function Navigator:switchOffHeadlights() 
  JavaWindow:setState(self.id, {"switchOffHeadlights"})   
end

function Navigator:retractLandingGears() 
  JavaWindow:setState(self.id, {"retractLandingGears"})   
end

function Navigator:extendLandingGears() 
  JavaWindow:setState(self.id, {"extendLandingGears"})   
end

function Navigator:isAnyLandingGearExtended() 
  return JavaWindow:get(self.id, {"isAnyLandingGearExtended"})   
end

function Navigator:setupControlMasterModeProperties(controlMasterModeId, displayName)
  JavaWindow:setState(self.id, {"setupControlMasterModeProperties", controlMasterModeId, displayName})    
end

function Navigator:getControlMasterModeId() 
  return JavaWindow:get(self.id, {"getControlMasterModeId"})   
end

function Navigator:setupAxisCommandProperties(axis, commandType)
  JavaWindow:setState(self.id, {"setupAxisCommandProperties", axis, commandType})    
end

function Navigator:setAxisCommandValue(axis, commandValue)
  JavaWindow:setState(self.id, {"setAxisCommandValue", axis, commandValue})    
end

function Navigator:getAxisCommandValue(axis) 
  return JavaWindow:get(self.id, {"getAxisCommandValue", axis})   
end

function Navigator:setEngineThrust(taglist, thrust)
  JavaWindow:setState(self.id, {"setEngineThrust", taglist, thrust})    
end

function Navigator:setEngineCommand(taglist, acceleration, angularAcceleration, keepForceCollinearity, keepTorqueCollinearity, 
                                    priority1SubTags, priority2SubTags, priority3SubTags, toleranceRatioToStopCommand)
  JavaWindow:setState(self.id, {"setEngineCommand",taglist, acceleration, angularAcceleration, keepForceCollinearity, keepTorqueCollinearity, 
                                    priority1SubTags, priority2SubTags, priority3SubTags, toleranceRatioToStopCommand})    
end                                    

function Navigator:getAtmosphereDensity() 
  return JavaWindow:get(self.id, {"getAtmosphereDensity"})   
end

function Navigator:getClosestPlanetInfluence() 
  return JavaWindow:get(self.id, {"getClosestPlanetInfluence"})   
end
